### PR TITLE
ci: dispatch rebuild event to algo-ai-training on main push

### PR DESCRIPTION
**Pull request by LeDuySon**
## Summary
- Adds a workflow that fires a `repository_dispatch` event (`ultralytics-updated`) to `SkillReal-LTD/algo-ai-training` whenever `main` is updated
- This triggers an automatic rebuild of the sagemaker training Docker image, which installs ultralytics from this fork

## Pre-requisite
- Add a secret `TRAINING_REPO_DISPATCH_TOKEN` to this repo (fine-grained PAT with **Actions: Read and write** on `algo-ai-training`)

## Test plan
- [ ] Confirm `TRAINING_REPO_DISPATCH_TOKEN` secret is configured
- [ ] Merge a test commit to `main` and verify the `ecr_training` workflow triggers in `algo-ai-training`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
🚦 Adds a new GitHub Action to dispatch a rebuild event to the external `algo-ai-training` repo on `main` pushes, but the PR also includes a broad set of unrelated training, metrics, W&B, segmentation, classification, config, docs, scripts, and test changes.

### 📊 Key Changes
- 🛠️ Adds `.github/workflows/trigger_training_rebuild.yaml` to send a `repository_dispatch` event to `SkillReal-LTD/algo-ai-training` after pushes to `main`.
- 📁 Updates `.gitignore` to exclude `/data`.
- 📦 Adds local dataset configs such as `configs/coco8.yaml`, `configs/coco8_pose.yaml`, and `configs/test_skillreal.yaml`.
- 🧪 Introduces many scripts and tests for fitness weighting, class weights, classification losses, boundary-aware segmentation metrics/losses, checkpoint debugging, GPU partition debugging, and local training workflows.
- 🧠 Extends core training and validation logic across detection, pose, segmentation, and classification, including custom `fitness_weight`, `class_weights`, boundary loss handling, extra metric logging, ArcFace/class-balanced/focal classification losses, and dataset channel overrides.
- 📊 Expands W&B integration and adds a custom `skillreal` callback for checkpoint syncing and metric logging.
- 📝 Adds multiple new documentation files covering best-model logging and W&B integration.
- ⚠️ Changes several defaults and integrations, including enabling W&B in settings by default and registering new callbacks.

### 🎯 Purpose & Impact
- 🔄 The stated CI goal is to automatically trigger downstream training image rebuilds whenever `main` is updated.
- 🚀 The broader code changes appear aimed at supporting internal or external training workflows with richer experiment control, logging, and debugging.
- ⚠️ Impact is high because this PR modifies many core training paths and defaults beyond CI, which may increase review complexity, regression risk, and maintenance burden.
- ✅ If split into smaller focused PRs, the CI workflow, docs, and training-feature additions would be easier to validate and merge safely.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @LeDuySon, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review triage, and an engineer will assist shortly. Since the PR title is CI-focused but the diff also includes broad training, metrics, callbacks, docs, scripts, and test changes, please double-check the scope and review the checklist below 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/cfg/default.yaml [hunk 1]
```diff
@@ -39,12 +39,22 @@ freeze: # (int | list, optional) freeze first N layers (int) or specific layer i
 multi_scale: 0.0 # (float) multi-scale range as a fraction of imgsz; sizes are rounded to stride multiples
 compile: False # (bool | str) enable torch.compile() backend='inductor'; True="default", False=off, or "default|reduce-overhead|max-autotune-no-cudagraphs"
 
+fitness_weight: [0.0, 0.9, 0.1, 0.0] # (list[float]) fitness weights for [P, R, mAP@0.5, mAP@0.5:0.95] for detection/OBB/classify tasks, or [box_P, box_R, box_mAP@0.5, box_mAP@0.5:0.95, pose_P/mask_P, pose_R/mask_R, pose_mAP@0.5/mask_mAP@0.5, pose_mAP@0.5:0.95/mask_mAP@0.5:0.95] for pose/segment tasks (8 values). Segment task additionally accepts 10 values: the 8-value layout followed by [mask_Dice, mask_BIoU] to weight boundary-quality metrics into fitness. Can be overridden for custom optimization targets.
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/default.yaml:42]: ⚠️ **HIGH**: This introduces a new global default that changes best-model selection from the longstanding mAP-weighted behavior to recall-weighted behavior for users who do not set `fitness_weight` explicitly. Because `args.fitness_weight` is now always populated, validators/metrics won't fall back to their legacy defaults anymore. That's a breaking behavioral change for all training runs and will skew benchmark/model-selection results unless it's an intentional project-wide migration.

File: ultralytics/data/build.py [hunk 2]
```diff
@@ -186,11 +195,32 @@ def _get_rank_indices(self) -> tuple[int, int]:
...
+            )
+
         return start_idx, end_idx
 
     def __iter__(self) -> Iterator:
         """Generate indices for this rank's contiguous chunk of the dataset."""
         start_idx, end_idx = self._get_rank_indices()
+
+        # Safety check: if no samples, return empty iterator to prevent infinite loop
+        if start_idx >= end_idx:
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py:221]: ⚠️ **HIGH**: Returning an empty iterator for ranks with no samples is unsafe under DDP. That rank will exhaust its dataloader immediately while other ranks keep stepping, which typically hangs on the next collective op/all-reduce. We need to fail fast with a clear error here, or pad/replicate samples so every rank still executes the same number of steps.

File: ultralytics/engine/trainer.py [hunk 3]
```diff
@@ -666,6 +796,13 @@ def get_dataset(self):
             LOGGER.info("Overriding class names with single class.")
             data["names"] = {0: "item"}
             data["nc"] = 1
+
+        # Override channels if specified in args
+        if hasattr(self.args, "channels") and self.args.channels is not None:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:801]: ⚠️ **HIGH**: With `channels: 3` now present in `default.yaml`, this branch will run for every training job and overwrite any dataset-defined `channels` value to 3. That breaks grayscale/multispectral datasets unless the user also passes `channels` on the CLI. We should only override the dataset value when `channels` was explicitly set by the user, not when it's just the config default.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR has useful additions, but I found three blocking behavioral regressions in core training code: the new default `fitness_weight` changes checkpoint selection globally, the new `channels` override now clobbers dataset-defined non-RGB channel counts, and the zero-sample sampler path can deadlock DDP instead of recovering safely. I’d address those before merging.

💬 Posted 3 inline comments

⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.
